File: src/read.ts
/** Reading shapefile components and zip archives back into GeoJSON. */
import { unzipSync } from 'fflate';

import { load } from './wasm.js';
import type { FeatureCollection, ReadOptions, ShapefileLayer } from './types.js';

/** The pieces of a shapefile to read. Only the `.shp` is required. */
export interface ShapefileSource {
  /** Contents of the `.shp`. The `.shx` is not needed. */
  shp: Uint8Array;
  /** Contents of the `.dbf`. Without it, features have empty properties. */
  dbf?: Uint8Array | undefined;
  /** Contents of the `.cpg`; decides how `.dbf` text is decoded. */
  cpg?: string | undefined;
  /** Contents of the `.prj`; surfaced as `wkt` on the result. */
  prj?: string | undefined;
}

/**
 * Reads shapefile components into a GeoJSON `FeatureCollection`.
 *
 * The `.shx` is not needed — it is only an index into the `.shp`, which this
 * reader walks sequentially.
 *
 * Shapefile polygons store their rings in one flat list with no nesting, so
 * holes are matched back to the ring that actually contains them (by point-in-
 * polygon test, choosing the smallest containing ring) and then rewound to
 * RFC 7946 winding order: exteriors counter-clockwise, holes clockwise.
 *
 * Single-part geometries come back as the simple GeoJSON type — `LineString`
 * rather than a one-element `MultiLineString`, and likewise for `Polygon`.
 *
 * `.dbf` text is decoded as UTF-8 unless the `.cpg` or `options.encoding` names
 * a legacy code page such as `cp1252`. dBase pads character columns to a fixed
 * width; that padding is always stripped.
 *
 * The WebAssembly module is instantiated on first use — no `init()` required.
 *
 * @param source The `.shp` bytes, and optionally the `.dbf`, `.cpg` and `.prj`.
 *   Without a `.dbf`, features come back with empty properties.
 * @param options Decoding settings. See {@link ReadOptions}.
 * @returns A `FeatureCollection`. When a `.prj` was supplied its text is carried
 *   on the non-standard `wkt` member, since GeoJSON has nowhere else to put it.
 *
 * @throws {Error} If the `.shp` is truncated, malformed, or holds a shape type
 *   this reader does not understand.
 *
 * @example
 * ```ts
 * const geojson = await readShapefile({
 *   shp: await readFile('roads.shp'),
 *   dbf: await readFile('roads.dbf'),
 *   prj: await readFile('roads.prj', 'utf8'),
 * });
 * ```
 *
 * @see {@link readShapefileZip} to read a whole archive at once.
 */
export async function readShapefile(
  source: ShapefileSource,
  options: ReadOptions = {},
): Promise<FeatureCollection> {
  const wasm = await load();

  return wasm.readShapefile(source.shp, source.dbf, {
    // An explicit option beats the .cpg, which is often absent or wrong.
    encoding: options.encoding ?? normalizeCpg(source.cpg),
    includeM: options.includeM ?? false,
    prj: source.prj,
  }) as FeatureCollection;
}

/**
 * Reads every shapefile inside a zip archive.
 *
 * Archives routinely hold more than one layer, and just as routinely nest them
 * in folders, so components are grouped by their full path minus extension.
 * `a/roads.shp` and `b/roads.shp` therefore stay separate layers.
 *
 * Each layer is decoded using its own `.cpg`, and carries its own `.prj`.
 * Directory entries, `__MACOSX/` resource forks and files with no matching
 * `.shp` are ignored.
 *
 * @param archive The zip file as bytes.
 * @param options Decoding settings. An explicit `encoding` here overrides the
 *   `.cpg` of every layer.
 * @returns One entry per layer, sorted by name.
 *
 * @throws {Error} If the archive cannot be unzipped, or holds no `.shp` at all.
 *
 * @example
 * ```ts
 * const layers = await readShapefileZip(bytes);
 * for (const layer of layers) {
 *   console.log(layer.name, layer.geojson.features.length, layer.prj);
 * }
 * ```
 *
 * @example Read an older archive written in a Windows code page
 * ```ts
 * const layers = await readShapefileZip(bytes, { encoding: 'cp1252' });
 * ```
 */
export async function readShapefileZip(
  archive: Uint8Array,
  options: ReadOptions = {},
): Promise<ShapefileLayer[]> {
  const files = unzipSync(archive);
  const decoder = new TextDecoder();

  // Group by path-without-extension so "roads.shp" and "roads.dbf" pair up, and
  // "a/roads.shp" stays distinct from "b/roads.shp".
  const groups = new Map<string, Record<string, Uint8Array>>();

  for (const [path, bytes] of Object.entries(files)) {
    // Directory entries and macOS resource forks.
    if (path.endsWith('/') || path.includes('__MACOSX/')) continue;

    const match = /^(.*)\.(shp|shx|dbf|prj|cpg)$/i.exec(path);
    if (!match?.[1] || !match[2]) continue;

    const stem = match[1];
    const group = groups.get(stem) ?? {};
    group[match[2].toLowerCase()] = bytes;
    groups.set(stem, group);
  }

  const layers: ShapefileLayer[] = [];

  for (const [stem, group] of groups) {
    // A .dbf or .prj with no .shp beside it is not a layer.
    if (!group.shp) continue;

    const cpg = group.cpg ? decoder.decode(group.cpg) : undefined;
    const prj = group.prj ? decoder.decode(group.prj).trim() : undefined;
    const encoding = options.encoding ?? normalizeCpg(cpg) ?? 'utf-8';

    layers.push({
      name: stem.split('/').pop() ?? stem,
      geojson: await readShapefile(
        { shp: group.shp, dbf: group.dbf, cpg, prj },
        { ...options, encoding },
      ),
      ...(prj ? { prj } : {}),
      encoding,
    });
  }

  if (layers.length === 0) {
    throw new Error('shapefile-wasm: the archive contains no .shp file.');
  }

  return layers.sort((a, b) => a.name.localeCompare(b.name));
}

/**
 * `.cpg` files hold a bare label with no agreed spelling — "UTF-8", "utf8",
 * "ISO-8859-1", or a raw code page number. Normalising here keeps the tolerance
 * in one place; the Rust side matches generously and falls back to UTF-8.
 */
function normalizeCpg(cpg?: string): string | undefined {
  if (!cpg) return undefined;
  // Strip a BOM and any trailing newline the file may carry.
  const cleaned = cpg.replace(/^﻿/, '').trim();
  return cleaned.length > 0 ? cleaned : undefined;
}

File: rust/read.rs
//! Reads shapefile components back into GeoJSON.
//!
//! Two things make this more than a mechanical transcription:
//!
//! * A shapefile polygon is a flat bag of rings with no nesting, so the holes
//!   have to be matched back to the ring that contains them before GeoJSON can
//!   describe them.
//! * A .dbf carries no encoding of its own. The caller passes whatever the
//!   companion .cpg said, and we fall back to UTF-8.

use std::io::Cursor;

use serde::Deserialize;
use serde_json::{Map, Number, Value};

use shapefile::record::traits::{HasM, HasXY, HasZ};
use shapefile::{Multipatch, Patch, Point, PointM, PointZ, Shape, ShapeReader, NO_DATA};

use crate::error::{Result, ShapefileError};

/// Caller-tunable knobs for reading.
/// `#[serde(default)]` fills absent fields from `Default`, so the derived impl
/// and the deserialised defaults are the same thing by construction.
#[derive(Debug, Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct ReadOptions {
    /// Character set of the .dbf, normally taken from the companion .cpg file.
    pub encoding: Option<String>,
    /// Emit shapefile M (measure) values as a trailing ordinate. GeoJSON has no
    /// notion of measures, so this is off by default.
    pub include_m: bool,
    /// Contents of the companion `.prj`, surfaced as `wkt` on the result.
    /// Passed through by the TypeScript layer; nothing here parses it.
    pub prj: Option<String>,
}

impl ReadOptions {
    /// `DynEncoding` is not exported by `dbase`, so the encoding is applied to
    /// the builder inside the match rather than returned from it.
    fn reader_builder(&self) -> dbase::ReaderBuilder {
        let builder = dbase::ReaderBuilder::new();

        let requested = self
            .encoding
            .as_deref()
            .unwrap_or("utf-8")
            .trim()
            .to_ascii_lowercase()
            .replace(['-', '_', ' '], "");

        // .cpg files are wildly inconsistent, so match generously.
        match requested.as_str() {
            "utf8" | "utf8bom" | "65001" | "" => builder.with_encoding(dbase::UnicodeLossy),
            "cp1252" | "windows1252" | "iso88591" | "latin1" | "ansi" | "1252" => {
                builder.with_encoding(yore::code_pages::CP1252)
            }
            "cp1250" | "windows1250" | "1250" => builder.with_encoding(yore::code_pages::CP1250),
            "cp1251" | "windows1251" | "1251" => builder.with_encoding(yore::code_pages::CP1251),
            "cp1253" | "windows1253" | "1253" => builder.with_encoding(yore::code_pages::CP1253),
            "cp1254" | "windows1254" | "1254" => builder.with_encoding(yore::code_pages::CP1254),
            "cp1255" | "windows1255" | "1255" => builder.with_encoding(yore::code_pages::CP1255),
            "cp1256" | "windows1256" | "1256" => builder.with_encoding(yore::code_pages::CP1256),
            "cp437" | "437" | "oem" => builder.with_encoding(yore::code_pages::CP437),
            "cp850" | "850" => builder.with_encoding(yore::code_pages::CP850),
            "cp852" | "852" => builder.with_encoding(yore::code_pages::CP852),
            "cp865" | "865" => builder.with_encoding(yore::code_pages::CP865),
            "cp866" | "866" => builder.with_encoding(yore::code_pages::CP866),
            "cp874" | "874" => builder.with_encoding(yore::code_pages::CP874),
            // An unrecognised label is far better handled as UTF-8 than as a
            // hard failure; worst case a few characters come back replaced.
            _ => builder.with_encoding(dbase::UnicodeLossy),
        }
    }
}

/// Reads a `.shp` (and optionally its `.dbf`) into a GeoJSON FeatureCollection.
pub fn read(shp: &[u8], dbf: Option<&[u8]>, options: &ReadOptions) -> Result<Value> {
    let reader = ShapeReader::new(Cursor::new(shp))?;
    let shapes = reader.read()?;

    // Shapes and records are paired positionally. Reading them separately (as
    // opposed to `shapefile::Reader`) means a truncated .dbf degrades to missing
    // attributes instead of failing the whole file.
    let records = match dbf {
        Some(bytes) => read_records(bytes, options)?,
        None => Vec::new(),
    };

    let mut features = Vec::with_capacity(shapes.len());
    for (index, shape) in shapes.iter().enumerate() {
        let mut feature = Map::new();
        feature.insert("type".into(), Value::String("Feature".into()));
        feature.insert(
            "geometry".into(),
            shape_to_geometry(shape, options.include_m, index)?,
        );
        feature.insert(
            "properties".into(),
            records
                .get(index)
                .cloned()
                .map(Value::Object)
                .unwrap_or(Value::Object(Map::new())),
        );
        features.push(Value::Object(feature));
    }

    let mut collection = Map::new();
    collection.insert("type".into(), Value::String("FeatureCollection".into()));
    collection.insert("features".into(), Value::Array(features));
    if let Some(prj) = &options.prj {
        // Non-standard, but the alternative is silently losing the projection.
        collection.insert("wkt".into(), Value::String(prj.clone()));
    }

    Ok(Value::Object(collection))
}

fn read_records(dbf: &[u8], options: &ReadOptions) -> Result<Vec<Map<String, Value>>> {
    let mut reader = options.reader_builder().build(Cursor::new(dbf))?;

    // `dbase::Record` is a hash map, so grab the declared field order first;
    // otherwise the GeoJSON properties come out shuffled.
    let field_names: Vec<String> = reader
        .fields()
        .iter()
        .map(|field| field.name().to_string())
        .filter(|name| name != "DeletionFlag")
        .collect();

    let records = reader.read()?;

    Ok(records
        .into_iter()
        .map(|record| {
            let mut properties = Map::new();
            for name in &field_names {
                let value = record.get(name).map(field_to_json).unwrap_or(Value::Null);
                properties.insert(name.clone(), value);
            }
            properties
        })
        .collect())
}

/// dBase pads character fields out to their declared width; `dbase` strips that
/// padding on read and offers no way to keep it, so values always arrive trimmed.
fn field_to_json(value: &dbase::FieldValue) -> Value {
    use dbase::FieldValue;

    let text = |value: &String| Value::String(value.clone());

    match value {
        FieldValue::Character(Some(value)) => text(value),
        FieldValue::Memo(value) => text(value),
        FieldValue::Character(None) => Value::Null,
        FieldValue::Numeric(Some(value)) => number(*value),
        FieldValue::Numeric(None) => Value::Null,
        FieldValue::Float(Some(value)) => number(*value as f64),
        FieldValue::Float(None) => Value::Null,
        FieldValue::Logical(Some(value)) => Value::Bool(*value),
        FieldValue::Logical(None) => Value::Null,
        FieldValue::Integer(value) => Value::Number(Number::from(*value)),
        FieldValue::Currency(value) | FieldValue::Double(value) => number(*value),
        FieldValue::Date(Some(date)) => Value::String(format!(
            "{:04}-{:02}-{:02}",
            date.year(),
            date.month(),
            date.day()
        )),
        FieldValue::Date(None) => Value::Null,
        FieldValue::DateTime(stamp) => {
            let date = stamp.date();
            let time = stamp.time();
            Value::String(format!(
                "{:04}-{:02}-{:02}T{:02}:{:02}:{:02}Z",
                date.year(),
                date.month(),
                date.day(),
                time.hours(),
                time.minutes(),
                time.seconds()
            ))
        }
    }
}

fn number(value: f64) -> Value {
    Number::from_f64(value)
        .map(Value::Number)
        .unwrap_or(Value::Null)
}

/// True when a measure ordinate carries the shapefile "no data" sentinel.
fn has_measure(m: f64) -> bool {
    m > NO_DATA && m.is_finite()
}

fn position_xy(point: &Point) -> Value {
    Value::Array(vec![number(point.x), number(point.y)])
}

fn position_m(point: &PointM, include_m: bool) -> Value {
    let mut ordinates = vec![number(point.x()), number(point.y())];
    if include_m && has_measure(point.m()) {
        ordinates.push(number(point.m()));
    }
    Value::Array(ordinates)
}

fn position_z(point: &PointZ, include_m: bool) -> Value {
    let mut ordinates = vec![number(point.x()), number(point.y()), number(point.z())];
    if include_m && has_measure(point.m()) {
        ordinates.push(number(point.m()));
    }
    Value::Array(ordinates)
}

fn geometry(kind: &str, coordinates: Value) -> Value {
    let mut object = Map::new();
    object.insert("type".into(), Value::String(kind.into()));
    object.insert("coordinates".into(), coordinates);
    Value::Object(object)
}

/// Wraps a list of parts as a LineString when there is exactly one, and a
/// MultiLineString otherwise — GeoJSON prefers the simpler type.
fn lines(parts: Vec<Value>) -> Value {
    if parts.len() == 1 {
        geometry("LineString", parts.into_iter().next().unwrap())
    } else {
        geometry("MultiLineString", Value::Array(parts))
    }
}

fn points(points: Vec<Value>) -> Value {
    geometry("MultiPoint", Value::Array(points))
}

fn shape_to_geometry(shape: &Shape, include_m: bool, index: usize) -> Result<Value> {
    let geometry = match shape {
        Shape::NullShape => Value::Null,

        Shape::Point(point) => geometry("Point", position_xy(point)),
        Shape::PointM(point) => geometry("Point", position_m(point, include_m)),
        Shape::PointZ(point) => geometry("Point", position_z(point, include_m)),

        Shape::Multipoint(shape) => points(shape.points().iter().map(position_xy).collect()),
        Shape::MultipointM(shape) => points(
            shape
                .points()
                .iter()
                .map(|point| position_m(point, include_m))
                .collect(),
        ),
        Shape::MultipointZ(shape) => points(
            shape
                .points()
                .iter()
                .map(|point| position_z(point, include_m))
                .collect(),
        ),

        Shape::Polyline(shape) => lines(
            shape
                .parts()
                .iter()
                .map(|part| Value::Array(part.iter().map(position_xy).collect()))
                .collect(),
        ),
        Shape::PolylineM(shape) => lines(
            shape
                .parts()
                .iter()
                .map(|part| Value::Array(part.iter().map(|p| position_m(p, include_m)).collect()))
                .collect(),
        ),
        Shape::PolylineZ(shape) => lines(
            shape
                .parts()
                .iter()
                .map(|part| Value::Array(part.iter().map(|p| position_z(p, include_m)).collect()))
                .collect(),
        ),

        Shape::Polygon(shape) => rings_to_geometry(
            shape
                .rings()
                .iter()
                .map(|ring| Ring::from_points(ring, |point| (point.x, point.y), position_xy))
                .collect(),
        ),
        Shape::PolygonM(shape) => rings_to_geometry(
            shape
                .rings()
                .iter()
                .map(|ring| {
                    Ring::from_points(
                        ring,
                        |point| (point.x(), point.y()),
                        |point| position_m(point, include_m),
                    )
                })
                .collect(),
        ),
        Shape::PolygonZ(shape) => rings_to_geometry(
            shape
                .rings()
                .iter()
                .map(|ring| {
                    Ring::from_points(
                        ring,
                        |point| (point.x(), point.y()),
                        |point| position_z(point, include_m),
                    )
                })
                .collect(),
        ),

        Shape::Multipatch(shape) => multipatch_to_geometry(shape, include_m, index)?,
    };

    Ok(geometry)
}

/// A polygon ring reduced to what the nesting logic needs: plain XY for the
/// containment maths, plus the already-formatted GeoJSON positions.
struct Ring {
    outer: bool,
    xy: Vec<(f64, f64)>,
    positions: Vec<Value>,
}

impl Ring {
    fn from_points<P, F, G>(ring: &shapefile::PolygonRing<P>, to_xy: F, to_position: G) -> Self
    where
        F: Fn(&P) -> (f64, f64),
        G: Fn(&P) -> Value,
    {
        let (outer, points) = match ring {
            shapefile::PolygonRing::Outer(points) => (true, points),
            shapefile::PolygonRing::Inner(points) => (false, points),
        };
        Self {
            outer,
            xy: points.iter().map(&to_xy).collect(),
            positions: points.iter().map(&to_position).collect(),
        }
    }

    /// Twice the signed area. Positive means counter-clockwise.
    fn signed_area(&self) -> f64 {
        let mut total = 0.0;
        for window in self.xy.windows(2) {
            let (x1, y1) = window[0];
            let (x2, y2) = window[1];
            total += (x2 - x1) * (y2 + y1);
        }
        -total
    }

    fn contains(&self, point: (f64, f64)) -> bool {
        // Standard ray casting; the ring is already closed.
        let (px, py) = point;
        let mut inside = false;
        for window in self.xy.windows(2) {
            let (x1, y1) = window[0];
            let (x2, y2) = window[1];
            if (y1 > py) != (y2 > py) {
                let slope = (x2 - x1) / (y2 - y1);
                if px < x1 + (py - y1) * slope {
                    inside = !inside;
                }
            }
        }
        inside
    }

    /// GeoJSON (RFC 7946) wants exteriors counter-clockwise and holes clockwise;
    /// shapefiles use the opposite convention.
    fn oriented(mut self, counter_clockwise: bool) -> Vec<Value> {
        let is_ccw = self.signed_area() > 0.0;
        if is_ccw != counter_clockwise {
            self.positions.reverse();
        }
        self.positions
    }
}

/// Rebuilds GeoJSON polygon nesting from a shapefile's flat ring list.
fn rings_to_geometry(rings: Vec<Ring>) -> Value {
    let (outers, inners): (Vec<Ring>, Vec<Ring>) = rings.into_iter().partition(|ring| ring.outer);

    if outers.is_empty() {
        // Malformed input: no ring was marked as an exterior. Treat each one as
        // its own polygon rather than dropping the geometry entirely.
        let polygons: Vec<Value> = inners
            .into_iter()
            .map(|ring| Value::Array(vec![Value::Array(ring.oriented(true))]))
            .collect();
        return finish_polygons(polygons);
    }

    // Each hole belongs to the smallest exterior ring that contains it. Using
    // the smallest matters when polygons are nested inside one another.
    let mut assignments: Vec<Vec<Value>> = outers.iter().map(|_| Vec::new()).collect();
    for inner in inners {
        let probe = match inner.xy.first() {
            Some(point) => *point,
            None => continue,
        };

        let best = outers
            .iter()
            .enumerate()
            .filter(|(_, outer)| outer.contains(probe))
            .min_by(|(_, a), (_, b)| {
                a.signed_area()
                    .abs()
                    .partial_cmp(&b.signed_area().abs())
                    .unwrap_or(std::cmp::Ordering::Equal)
            })
            .map(|(index, _)| index);

        // An unmatched hole is a data error; parking it on the first ring keeps
        // the output valid GeoJSON.
        let target = best.unwrap_or(0);
        assignments[target].push(Value::Array(inner.oriented(false)));
    }

    let polygons: Vec<Value> = outers
        .into_iter()
        .zip(assignments)
        .map(|(outer, holes)| {
            let mut ring_list = vec![Value::Array(outer.oriented(true))];
            ring_list.extend(holes);
            Value::Array(ring_list)
        })
        .collect();

    finish_polygons(polygons)
}

fn finish_polygons(polygons: Vec<Value>) -> Value {
    if polygons.len() == 1 {
        geometry("Polygon", polygons.into_iter().next().unwrap())
    } else {
        geometry("MultiPolygon", Value::Array(polygons))
    }
}

/// Multipatch has no GeoJSON equivalent. Triangle strips and fans are expanded
/// into individual triangles so the surface survives as a MultiPolygon; the
/// alternative is discarding the geometry.
fn multipatch_to_geometry(shape: &Multipatch, include_m: bool, index: usize) -> Result<Value> {
    let mut polygons: Vec<Value> = Vec::new();
    let mut pending_outer: Option<Vec<Value>> = None;

    let close = |points: &[PointZ]| -> Vec<Value> {
        let mut positions: Vec<Value> = points.iter().map(|p| position_z(p, include_m)).collect();
        if positions.len() > 2 && positions.first() != positions.last() {
            positions.push(positions[0].clone());
        }
        positions
    };

    for patch in shape.patches() {
        match patch {
            Patch::TriangleStrip(points) => {
                for triangle in points.windows(3) {
                    polygons.push(Value::Array(vec![Value::Array(close(triangle))]));
                }
            }
            Patch::TriangleFan(points) => {
                if points.len() >= 3 {
                    for pair in points[1..].windows(2) {
                        let triangle = [points[0], pair[0], pair[1]];
                        polygons.push(Value::Array(vec![Value::Array(close(&triangle))]));
                    }
                }
            }
            Patch::OuterRing(points) => {
                if let Some(previous) = pending_outer.take() {
                    polygons.push(Value::Array(vec![Value::Array(previous)]));
                }
                pending_outer = Some(close(points));
            }
            Patch::InnerRing(points) => {
                // An inner ring belongs to the outer ring that preceded it.
                let hole = Value::Array(close(points));
                match polygons.last_mut() {
                    Some(Value::Array(rings)) if pending_outer.is_none() => rings.push(hole),
                    _ => {
                        if let Some(outer) = pending_outer.take() {
                            polygons.push(Value::Array(vec![Value::Array(outer), hole]));
                        }
                    }
                }
            }
            Patch::FirstRing(points) | Patch::Ring(points) => {
                if let Some(previous) = pending_outer.take() {
                    polygons.push(Value::Array(vec![Value::Array(previous)]));
                }
                polygons.push(Value::Array(vec![Value::Array(close(points))]));
            }
        }
    }

    if let Some(outer) = pending_outer.take() {
        polygons.push(Value::Array(vec![Value::Array(outer)]));
    }

    if polygons.is_empty() {
        return Err(ShapefileError::Feature {
            index,
            message: "multipatch contained no renderable patches".into(),
        });
    }

    Ok(finish_polygons(polygons))
}
